server/hotrod/src/main/java/org/infinispan/server/hotrod/ClientListenerRegistry.java

```diff
@@ -388,16 +390,38 @@ void sendEvent(byte[] key, byte[] value, long dataVersion, CacheEntryEvent event
          if (isTrace)
             log.tracef("Queue event %s, before queuing event queue size is %d", remoteEvent, eventQueue.size());
 
-         boolean waitingForFlush = !ch.isWritable();
-         try {
-            eventQueue.put(remoteEvent);
-         } catch (InterruptedException e) {
-            throw new CacheException(e);
-         }
+         EventLoop loop = ch.eventLoop();
+         // We shouldn't be in the event loop, but just in case we can't get stuck blocking on putting into the queue
+         // so we offer and try to catch up on events if possible
+         if (loop.inEventLoop()) {
+            boolean offered = eventQueue.offer(remoteEvent);
+            while (!offered) {
+               // If the event queue is full, we have to try to write some events to free up space since we are in the
+               // event loop and no other thread can drain this queue but us
+               writeEventsIfPossible();
+               // We again try to offer, but if we weren't able to write any events this will not offer - so we
+               // have to wait for the client to catch up to us - we put a little wait to not cause CPU to spin
+               try {
+                  offered = eventQueue.offer(remoteEvent, 1, TimeUnit.MILLISECONDS);
+               } catch (InterruptedException e) {
+                  throw new CacheException(e);
+               }
+            }
+            if (ch.isWritable()) {
+               // Now that the entry is finally in eventQueue, submit the events to be written at a later point
+               loop.submit(writeEventsIfPossible);
+            }
+         } else {
+            try {
+               eventQueue.put(remoteEvent);
+            } catch (InterruptedException e) {
+               throw new CacheException(e);
+            }
 
-         if (!waitingForFlush) {
-            // Make sure we write any event in main event loop
-            ch.eventLoop().submit(writeEventsIfPossible);
+            if (ch.isWritable()) {
+               // Make sure we write any event in main event loop
+               loop.submit(writeEventsIfPossible);
```

Comment:
This is the same in both branches. The comment is different, but this one is better ;)